### ISSUE TITLE
JSON return type for Handlers?

### ISSUE THREAD

**Issue by MaxGabriel** [yesodweb/yesod#1481]
So I was thinking about this blog post recently http://tech.frontrowed.com/2015/12/21/servant-style-handlers-for-yesod/, that uses `UndecidableInstances` to let you specify the return type of your Handlers that return JSON as `Handler <RecordName>` rather than `Handler Value`

Why not just have something like this?

```haskell
data JSONResponse a = ToJSON a => JSONResponse a

instance ToContent (JSONResponse a) where
  toContent (JSONResponse a) = toContent $ toEncoding a

instance ToTypedContent (JSONResponse a) where
  toTypedContent = TypedContent typeJson . toContent
```

Then handlers are typed as:

```haskell
postSignupR :: Handler (JSONResponse CreateUserResponse)
```

and the function returns: 
```haskell
return $ JSONResponse $ CreateUserResponse userId
```

Slightly more overhead than the blog post version. I don’t know if `UndecidableInstances` is actually worth avoiding though. Thoughts? I think endorsing one or the other of these options would be good for Yesod to do—it seems much better than returning `Value`.

**Comment by MaxGabriel** [comment]
cc @gregwebs since you authored that blog post

**Comment by gregwebs** [comment]
Looks good

**Comment by StevenXL** [comment]
Closed by #1592
